addon/-private/set-deep.ts

```diff
@@ -0,0 +1,68 @@
+import Change from './change';
+/**
+ * TODO: consider
+ * https://github.com/emberjs/ember.js/blob/822452c4432620fc67a777aba3b150098fd6812d/packages/%40ember/-internals/metal/lib/property_set.ts
+ *
+ * @method setDeep
+ * @param target
+ * @param path
+ * @param value
+ */
+export default function setDeep(target: any, path: string, value: unknown): any {
+  const keys = split(path).filter(isValidKey);
+  let orig = target;
+
+  if (keys.length === 1) {
+    target[path] = value;
+    return target;
+  }
+
+  for (let i = 0; i < keys.length; i++) {
+    let prop = keys[i];
+
+    if (!isObject(target[prop])) {
+      target[prop] = {};
+    } else if (isObject(target[prop]) && target[prop] instanceof Change) {
+      // we don't want to merge new changes with a Change instance higher up in the obj tree
+      target[prop] = {};
+    }
+
+    // last iteration
+    if (i === keys.length - 1) {
+      result(target, prop, value);
```

Comment:
I think this could be refactored to extract the last iteration check out of the loop.
E.g. loop until `keys.length - 1` and always do the result call afterwards.